mlir/lib/IR/AffineExpr.cpp

```diff
@@ -237,6 +239,135 @@ unsigned AffineDimExpr::getPosition() const {
   return static_cast<ImplType *>(expr)->position;
 }
 
+/// Returns true if this expression is made out of only symbols.
+static bool isSymbolic(AffineExpr expr) {
+  if (expr.getKind() == AffineExprKind::SymbolId)
+    return true;
+  return false;
+}
+
+/// Returns true if the expression is divisible by the given symbol.
+static bool isDivisibleBySymbol(AffineExpr expr, unsigned symbolPos) {
+  switch (expr.getKind()) {
+  case AffineExprKind::Constant:
+    return false;
+  case AffineExprKind::DimId:
+    return false;
+  case AffineExprKind::SymbolId: {
+    AffineSymbolExpr sym = expr.dyn_cast<AffineSymbolExpr>();
+    if (sym.getPosition() == symbolPos)
+      return true;
+    return false;
+  }
+  case AffineExprKind::Add: {
+    AffineBinaryOpExpr binaryExpr = expr.cast<AffineBinaryOpExpr>();
+    return isDivisibleBySymbol(binaryExpr.getLHS(), symbolPos) &&
+           isDivisibleBySymbol(binaryExpr.getRHS(), symbolPos);
+  }
+  case AffineExprKind::Mul: {
+    AffineBinaryOpExpr binaryExpr = expr.cast<AffineBinaryOpExpr>();
+    return isDivisibleBySymbol(binaryExpr.getLHS(), symbolPos) ||
+           isDivisibleBySymbol(binaryExpr.getRHS(), symbolPos);
+  }
+  case AffineExprKind::FloorDiv:
+  case AffineExprKind::CeilDiv:
+  case AffineExprKind::Mod: {
+    AffineBinaryOpExpr binaryExpr = expr.cast<AffineBinaryOpExpr>();
+    return isDivisibleBySymbol(binaryExpr.getLHS(), symbolPos);
+  }
+  }
+  llvm_unreachable("Unknown AffineExpr");
+}
+
+/// Returns a affine expression by dividing given affine expression by given
+/// symbol.
```

Comment:
Use all the width.

Reply:
In comments can we ignore clang-format? You are saying to put "symbol." in the same line right?

Reply:
Oh, sorry, I thought it would within 80 columns. Please always follow clang-format.